pkg/pmem-csi-driver/pmem-csi-driver.go

```diff
@@ -177,9 +177,8 @@ func (pmemd *pmemDriver) Run() error {
 					return err
 				}
 			}
-			if err := pmemd.registerNodeController(); err != nil {
-				return err
-			}
+			go pmemd.registerNodeController()
```

Comment:
This allows the driver to go into a "ready" state (= accepting CSI calls) before it has successfully registered. Is that what we want? We don't have readiness probes at the moment, but if we had something based on the driver responding to CSI requests, then this would incorrectly report "ready" when in reality the driver is still starting up.

It's just a gut feeling, but I think I prefer calling `pmemd.registerNodeController()` in blocking mode.

Reply:
Also, what happens if registration fails? Does the driver now keep running despite the failure?

Reply:
> Also, what happens if registration fails? Does the driver now keep running despite the failure?

Yes, i would keep retry. In reality, the only valid path is missing arguments, both are validated at client side.

But, i agree, it should fail on real error case. Now i fixed this.

Reply:
> This allows the driver to go into a "ready" state (= accepting CSI calls) before it has successfully registered. Is that what we want? We don't have readiness probes at the moment, but if we had something based on the driver responding to CSI requests, then this would incorrectly report "ready" when in reality the driver is still starting up.
> 
> It's just a gut feeling, but I think I prefer calling `pmemd.registerNodeController()` in blocking mode.

Moved this to blocking call.

Reply:
I had a closer look at how `pmemd.registerNodeController` is called. Making it a blocking call is not enough, it also needs to be moved above the `s.Start` calls, otherwise the gRPC server is already accepting incoming requests while the registration is still pending.

Reply:
@pohly I feel, registering is not a hard requirement from driver at node, and it might not effect the 'readiness' state of node driver. It still can server valid node operations coming from kubelet and  driver registrar. I couldn't find a failure case here.

Can you help me understand if i miss something here.

Reply:
The failure is that the system as a whole isn't operational when nodes don't register themselves. Consider how an admin currently checks whether the pmem-csi driver deployment is working. Most likely they'll just check that all pods (including the sidecars) are running and "kubectl describe" doesn't show failures.

But if the node currently fails to register, all pods are running and yet provisioning will fail or (worse) silently ignore nodes it doesn't know about.

An admin would have to grep the logs to determine whether everything is alright. That's so unusual, I doubt people will do it, even if we put it in the documentation.

Reply:
Sorry if my earlier comment sounded like "node registration failure can be ignored". That was not my intention. It was in the context of your suggestion to move `pmemd.registerNodeController()`  before starting gRPC {identity, controller, node}servers.  If we start controller after registration, Any requests to this node controller made by Master from its registration handler might fail. So, we should start ControllerServer by the time we call `registry.RegisterController()`

Starting identity and node servers prior to registration, which allows to register the driver at node with kubelet, prior to registration. Does this leads to issues? I couldn't see any.

And after reading your comments, i felt like we should not keep retry to connect forever, fail after may be n-tries?

Reply:
> If we start controller after registration, Any requests to this node controller made by Master from its registration handler might fail.

No, they don't. The listening socket is open, so connections from master get accepted and queued. The response will be generated as soon as the driver continues from "registering myself" to "serving requests".

> Starting identity and node servers prior to registration, which allows to register the driver at node with kubelet, prior to registration. Does this leads to issues? I couldn't see any.

My idea was to not create the Unix domain `csi.sock` and thus indicating to sidecars that the driver is not ready yet. However, that alone doesn't help an admin, because the sidecars will be running. What we need is a "readiness probe" sidecar. I'll ask in the CSI standup meeting why they implemented a "liveness probe" sidecar, but not a "readiness probe". IMHO "readiness" is more important.

> And after reading your comments, i felt like we should not keep retry to connect forever, fail after may be n-tries?

I still think trying to connect forever is the right approach, in particular if an admin can tell that the driver is still initializing (see "readiness" above).

Reply:
> No, they don't. The listening socket is open,

Listening Endpoint socket is only open in `Start()` call, without calling that no listening socket and if Master controller tries to connect to "Endpoint" we asked to register from its registration handler might fail?

Reply:
Forget what I said then. I was thinking of "Start" as "start gRPC server" because that is how it is done elsewhere (if I remember correctly), but here it is indeed different.

We probably need to revisit this entire "bring up pmem-csi" topic sometime on the future when considering what it means for pmem-csi to be "ready" and what different failure scenarios could be.

Reply:
Ok, i did what the best we can do with current code:
- Start ControllerServer
- Initiate Register, exit if fail.
- Start identity and node servers.